v2/bundle/x509bundle/bundle.go

```diff
@@ -1,65 +1,182 @@
+		}
+
+		cert, err := x509.ParseCertificate(pemBlock.Bytes)
+		if err != nil {
+			return nil, fmt.Errorf("cannot parse certificate: %v", err)
+		}
+		bundle.AddX509Root(cert)
+	}
+	return bundle, nil
 }
 
 // TrustDomain returns the trust domain of the bundle.
 func (b *Bundle) TrustDomain() spiffeid.TrustDomain {
-	panic("not implemented")
+	b.tdMtx.RLock()
+	defer b.tdMtx.RUnlock()
+	return b.trustDomain
 }
 
 // X509Roots returns the X.509 roots in the bundle.
 func (b *Bundle) X509Roots() []*x509.Certificate {
-	panic("not implemented")
+	b.rootsMtx.RLock()
+	defer b.rootsMtx.RUnlock()
+	return b.roots
 }
 
 // AddX509Root adds an X.509 root to the bundle. If the root already
 // exists in the bundle, the contents of the bundle will remain unchanged.
-func (b *Bundle) AddX509Root(*x509.Certificate) {
-	panic("not implemented")
+func (b *Bundle) AddX509Root(root *x509.Certificate) {
+	b.rootsMtx.Lock()
+	defer b.rootsMtx.Unlock()
+
+	for _, r := range b.roots {
+		if reflect.DeepEqual(r, root) {
+			return
+		}
+	}
+
+	b.roots = append(b.roots, root)
 }
 
 // RemoveX509Root removes an X.509 root to the bundle.
-func (b *Bundle) RemoveX509Root(*x509.Certificate) {
-	panic("not implemented")
+func (b *Bundle) RemoveX509Root(root *x509.Certificate) {
+	b.rootsMtx.Lock()
+	defer b.rootsMtx.Unlock()
+
+	for i, r := range b.roots {
+		if reflect.DeepEqual(r, root) {
+			// Error can be safely ignored since the index comes from
+			// iterating on the root certificates slice
+			b.roots, _ = removeElement(b.roots, i)
+			return
+		}
+	}
+}
+
+// HasX509Root checks if the given X.509 root exists in the bundle
+func (b *Bundle) HasX509Root(root *x509.Certificate) bool {
+	b.rootsMtx.RLock()
+	defer b.rootsMtx.RUnlock()
+
+	for _, r := range b.roots {
+		if reflect.DeepEqual(r, root) {
+			return true
+		}
+	}
+	return false
 }
 
 // Marshal marshals the X.509 bundle into PEM-encoded certificate blocks.
 func (b *Bundle) Marshal() ([]byte, error) {
-	panic("not implemented")
+	b.rootsMtx.RLock()
+	defer b.rootsMtx.RUnlock()
+
+	var buf bytes.Buffer
+	for _, root := range b.roots {
+		err := pem.Encode(&buf, &pem.Block{
+			Type:  certType,
+			Bytes: root.Raw,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("unable to encode root certificate: %v", err)
+		}
+	}
+
+	return buf.Bytes(), nil
 }
 
 // GetX509BundleForTrustDomain returns the X.509 bundle for the given trust
 // domain. It implements the Source interface. It will fail if
 // called with a trust domain other than the one the bundle belongs to.
 func (b *Bundle) GetX509BundleForTrustDomain(trustDomain spiffeid.TrustDomain) (*Bundle, error) {
-	panic("not implemented")
+	b.tdMtx.RLock()
+	defer b.tdMtx.RUnlock()
+
+	if b.trustDomain != trustDomain {
+		return nil, errors.New("wrong td")
+	}
+
+	return b, nil
+}
+
+// removeElement removes an element from slice, slice order is not preserved.
+func removeElement(slice []*x509.Certificate, index int) ([]*x509.Certificate, error) {
```

Comment:
This seems to be only used in one spot, and that spot has a valid index.... I'd recommend either moving the body into RemoveX509Root or getting rid of the bounds checking since it is statically valid.

Reply:
By putting this in a separate function, my intention was to have better readability. But then I  thought that this method may be used in other places in the future and got concerned about index checking.

I believe I rather move the body into RemoveX509Root instead of getting rid of the bounds checking. Just in case.